compiler/rustc_metadata/src/dynamic_lib.rs

```diff
@@ -70,13 +70,12 @@ mod dl {
         use std::sync::{Mutex, MutexGuard};
 
         pub fn lock() -> MutexGuard<'static, Guard> {
-            static LOCK: SyncLazy<Mutex<Guard>> = SyncLazy::new(|| Mutex::new(Guard { _priv: () }));
+            static LOCK: SyncLazy<Mutex<Guard>> = SyncLazy::new(|| Mutex::new(Guard));
             LOCK.lock().unwrap()
         }
 
-        pub struct Guard {
-            _priv: (),
-        }
+        #[non_exhaustive]
+        pub struct Guard;
```

Comment:
Doesn't this permit construction of `Guard` by users who can't access Guard's fields? Like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1a1787b0c620b6ae21a4a2ad4f4b2fd9).

Reply:
Inside of a crate, `#[non_exhaustive]` has no effect. To external users, you can effectively pretend there's a hidden private field. There's an RFC and presumably a fair amount of documentation about what this feature is.

Reply:
That's true. My concern still applies equally much. `rustc_metadata` has a bunch of stuff in it, not just the dynamic loading utilities, and so there's still value in not (easily) allowing other code in `rustc_metadata` to accidentally shoot their foot off.

Reply:
I suppose, but I don't particularly see that being an issue given the code reviews done on any PR.